File: sealed_app/lib/remote/ohttp/binary_http.dart
import 'dart:typed_data';

/// Binary HTTP message framing (RFC 9292).
///
/// Encodes HTTP requests/responses into a binary format suitable for
/// OHTTP encapsulation.
class BinaryHttp {
  /// Encode an HTTP request into Binary HTTP format (Known-Length Request).
  ///
  /// Format:
  ///   Framing Indicator (1 byte, 0x00 for known-length request)
  ///   Method Length (variable-length int) | Method
  ///   Scheme Length (variable-length int) | Scheme
  ///   Authority Length (variable-length int) | Authority
  ///   Path Length (variable-length int) | Path
  ///   Headers (length-prefixed block)
  ///   Content (length-prefixed block)
  static Uint8List encodeRequest({
    required String method,
    required Uri uri,
    Map<String, String> headers = const {},
    Uint8List? body,
  }) {
    final builder = BytesBuilder();

    // Framing indicator: 0x00 = Known-Length Request
    builder.addByte(0x00);

    // Method
    _writeVarLenString(builder, method);

    // Scheme
    _writeVarLenString(builder, uri.scheme);

    // Authority (host:port)
    final authority = uri.hasPort && uri.port != 443 && uri.port != 80
        ? '${uri.host}:${uri.port}'
        : uri.host;
    _writeVarLenString(builder, authority);

    // Path (including query)
    final path = uri.hasQuery ? '${uri.path}?${uri.query}' : uri.path;
    _writeVarLenString(builder, path.isEmpty ? '/' : path);

    // Headers as a length-prefixed block
    final headerBytes = _encodeHeaders(headers);
    _writeVarLenBytes(builder, headerBytes);

    // Content (body)
    _writeVarLenBytes(builder, body ?? Uint8List(0));

    // Known-Length Trailers (empty)
    _writeVarInt(builder, 0);

    return Uint8List.fromList(builder.toBytes());
  }

  /// Decode a Binary HTTP response (Known-Length Response).
  ///
  /// Returns status code, headers, and body.
  static BinaryHttpResponse decodeResponse(Uint8List data) {
    int offset = 0;

    // Framing indicator: should be 0x01 for known-length response
    // But some implementations use informational responses (1xx) first.
    // Skip any informational responses.
    while (offset < data.length) {
      final framing = data[offset];
      offset += 1;

      if (framing == 0x01) {
        // Known-Length Response — read status and fields
        final (statusCode, o1) = _readVarInt(data, offset);
        offset = o1;

        // Headers
        final (headerBytes, o2) = _readVarLenBytes(data, offset);
        offset = o2;
        final headers = _decodeHeaders(headerBytes);

        // Content
        final (body, o3) = _readVarLenBytes(data, offset);
        offset = o3;

        return BinaryHttpResponse(
          statusCode: statusCode,
          headers: headers,
          body: body,
        );
      } else if (framing == 0x00) {
        // This is a request, not a response
        throw const FormatException('Expected response, got request framing');
      } else {
        // Unknown framing, skip
        throw FormatException('Unknown framing indicator: $framing');
      }
    }

    throw const FormatException('Empty Binary HTTP response');
  }

  // =========================================================================
  // Private helpers
  // =========================================================================

  static void _writeVarLenString(BytesBuilder builder, String value) {
    final bytes = Uint8List.fromList(value.codeUnits);
    _writeVarInt(builder, bytes.length);
    builder.add(bytes);
  }

  static void _writeVarLenBytes(BytesBuilder builder, Uint8List bytes) {
    _writeVarInt(builder, bytes.length);
    builder.add(bytes);
  }

  /// Write a variable-length integer (QUIC encoding).
  static void _writeVarInt(BytesBuilder builder, int value) {
    if (value < 0x40) {
      builder.addByte(value);
    } else if (value < 0x4000) {
      builder.addByte(0x40 | (value >> 8));
      builder.addByte(value & 0xFF);
    } else if (value < 0x40000000) {
      builder.addByte(0x80 | (value >> 24));
      builder.addByte((value >> 16) & 0xFF);
      builder.addByte((value >> 8) & 0xFF);
      builder.addByte(value & 0xFF);
    } else {
      throw ArgumentError('Value too large for variable-length int: $value');
    }
  }

  /// Read a variable-length integer (QUIC encoding).
  static (int, int) _readVarInt(Uint8List data, int offset) {
    if (offset >= data.length) {
      throw const FormatException('Unexpected end of data reading varint');
    }
    final first = data[offset];
    final prefix = first >> 6;

    switch (prefix) {
      case 0: // 1 byte
        return (first, offset + 1);
      case 1: // 2 bytes
        if (offset + 2 > data.length) {
          throw const FormatException('Truncated 2-byte varint');
        }
        final value = ((first & 0x3F) << 8) | data[offset + 1];
        return (value, offset + 2);
      case 2: // 4 bytes
        if (offset + 4 > data.length) {
          throw const FormatException('Truncated 4-byte varint');
        }
        final value =
            ((first & 0x3F) << 24) |
            (data[offset + 1] << 16) |
            (data[offset + 2] << 8) |
            data[offset + 3];
        return (value, offset + 4);
      default: // 8 bytes - not supported for our use case
        throw const FormatException('8-byte varint not supported');
    }
  }

  static (Uint8List, int) _readVarLenBytes(Uint8List data, int offset) {
    final (length, newOffset) = _readVarInt(data, offset);
    if (newOffset + length > data.length) {
      throw FormatException(
        'Truncated field: expected $length bytes at offset $newOffset',
      );
    }
    return (
      Uint8List.fromList(data.sublist(newOffset, newOffset + length)),
      newOffset + length,
    );
  }

  static Uint8List _encodeHeaders(Map<String, String> headers) {
    final builder = BytesBuilder();
    for (final entry in headers.entries) {
      final name = entry.key.toLowerCase().codeUnits;
      final value = entry.value.codeUnits;
      _writeVarInt(builder, name.length);
      builder.add(name);
      _writeVarInt(builder, value.length);
      builder.add(value);
    }
    return Uint8List.fromList(builder.toBytes());
  }

  static Map<String, String> _decodeHeaders(Uint8List data) {
    final headers = <String, String>{};
    int offset = 0;
    while (offset < data.length) {
      final (nameBytes, o1) = _readVarLenBytes(data, offset);
      offset = o1;
      final (valueBytes, o2) = _readVarLenBytes(data, offset);
      offset = o2;
      headers[String.fromCharCodes(nameBytes)] = String.fromCharCodes(
        valueBytes,
      );
    }
    return headers;
  }
}

/// Decoded Binary HTTP response.
class BinaryHttpResponse {
  final int statusCode;
  final Map<String, String> headers;
  final Uint8List body;

  BinaryHttpResponse({
    required this.statusCode,
    required this.headers,
    required this.body,
  });
}

File: sealed-indexer/src/notifications/binary-http.ts
/**
 * Binary HTTP message framing (RFC 9292).
 *
 * Encodes HTTP requests / decodes HTTP responses into the wire format expected
 * inside an OHTTP encapsulated request. Ported 1:1 from the Flutter client at
 * `sealed_app/lib/remote/ohttp/binary_http.dart` so both sides agree on the
 * exact byte layout.
 *
 * Known-Length Request format (framing indicator 0x00):
 *   varint(method) || method
 *   varint(scheme) || scheme
 *   varint(authority) || authority
 *   varint(path)     || path
 *   varint(headersLen) || headersBlock        (each header: varint(nameLen) name varint(valueLen) value)
 *   varint(contentLen) || content
 *   varint(0)                                  (empty trailers)
 */

export interface BinaryHttpRequest {
  method: string;
  url: string;                      // absolute URL — scheme/authority/path extracted
  headers: Record<string, string>;
  body: Uint8Array;
}

export interface BinaryHttpResponse {
  statusCode: number;
  headers: Record<string, string>;
  body: Uint8Array;
}

export function encodeRequest(req: BinaryHttpRequest): Uint8Array {
  const u = new URL(req.url);
  const out: number[] = [];

  // Framing indicator: 0x00 = Known-Length Request
  out.push(0x00);

  writeVarLenString(out, req.method);
  writeVarLenString(out, u.protocol.replace(/:$/, ''));   // "https" not "https:"

  // Authority: include port only if non-default
  const port = u.port;
  const isDefaultPort =
    (u.protocol === 'https:' && (port === '' || port === '443')) ||
    (u.protocol === 'http:' && (port === '' || port === '80'));
  const authority = isDefaultPort ? u.hostname : `${u.hostname}:${port}`;
  writeVarLenString(out, authority);

  // Path (with query if present)
  const path = (u.pathname || '/') + (u.search || '');
  writeVarLenString(out, path);

  // Headers block — length-prefixed
  const headerBytes = encodeHeaders(req.headers);
  writeVarInt(out, headerBytes.length);
  for (const b of headerBytes) out.push(b);

  // Content — length-prefixed
  writeVarInt(out, req.body.length);
  for (const b of req.body) out.push(b);

  // Empty trailers
  writeVarInt(out, 0);

  return new Uint8Array(out);
}

export function decodeResponse(data: Uint8Array): BinaryHttpResponse {
  let offset = 0;
  if (data.length === 0) {
    throw new Error('bhttp: empty response');
  }

  // Skip any informational (1xx) framings; expect final 0x01 known-length response.
  while (offset < data.length) {
    const framing = data[offset++];
    if (framing === 0x01) {
      const [statusCode, o1] = readVarInt(data, offset);
      offset = o1;

      const [headerBytes, o2] = readVarLenBytes(data, offset);
      offset = o2;
      const headers = decodeHeaders(headerBytes);

      const [body] = readVarLenBytes(data, offset);
      return { statusCode, headers, body };
    }
    if (framing === 0x00) {
      throw new Error('bhttp: expected response, got request framing');
    }
    throw new Error(`bhttp: unknown framing indicator 0x${framing.toString(16)}`);
  }
  throw new Error('bhttp: response truncated');
}

// ---------------------------------------------------------------------------
// Internals — QUIC-style varints (RFC 9000 §16) and length-prefixed fields.
// ---------------------------------------------------------------------------

function writeVarLenString(out: number[], value: string): void {
  const bytes = utf8(value);
  writeVarInt(out, bytes.length);
  for (const b of bytes) out.push(b);
}

function writeVarInt(out: number[], value: number): void {
  if (value < 0x40) {
    out.push(value);
  } else if (value < 0x4000) {
    out.push(0x40 | (value >> 8));
    out.push(value & 0xff);
  } else if (value < 0x40000000) {
    out.push(0x80 | (value >> 24));
    out.push((value >> 16) & 0xff);
    out.push((value >> 8) & 0xff);
    out.push(value & 0xff);
  } else {
    throw new Error(`varint: value too large (${value})`);
  }
}

function readVarInt(data: Uint8Array, offset: number): [number, number] {
  if (offset >= data.length) throw new Error('varint: unexpected end of data');
  const first = data[offset];
  const prefix = first >> 6;
  switch (prefix) {
    case 0:
      return [first, offset + 1];
    case 1:
      if (offset + 2 > data.length) throw new Error('varint: truncated 2-byte');
      return [((first & 0x3f) << 8) | data[offset + 1], offset + 2];
    case 2:
      if (offset + 4 > data.length) throw new Error('varint: truncated 4-byte');
      return [
        ((first & 0x3f) << 24) |
          (data[offset + 1] << 16) |
          (data[offset + 2] << 8) |
          data[offset + 3],
        offset + 4,
      ];
    default:
      throw new Error('varint: 8-byte not supported');
  }
}

function readVarLenBytes(data: Uint8Array, offset: number): [Uint8Array, number] {
  const [len, newOffset] = readVarInt(data, offset);
  if (newOffset + len > data.length) {
    throw new Error(`bhttp: truncated field (need ${len} bytes at ${newOffset})`);
  }
  return [data.subarray(newOffset, newOffset + len), newOffset + len];
}

function encodeHeaders(headers: Record<string, string>): Uint8Array {
  const out: number[] = [];
  for (const [k, v] of Object.entries(headers)) {
    const name = utf8(k.toLowerCase());
    const value = utf8(v);
    writeVarInt(out, name.length);
    for (const b of name) out.push(b);
    writeVarInt(out, value.length);
    for (const b of value) out.push(b);
  }
  return new Uint8Array(out);
}

function decodeHeaders(data: Uint8Array): Record<string, string> {
  const headers: Record<string, string> = {};
  let offset = 0;
  while (offset < data.length) {
    const [nameBytes, o1] = readVarLenBytes(data, offset);
    offset = o1;
    const [valueBytes, o2] = readVarLenBytes(data, offset);
    offset = o2;
    headers[utf8Decode(nameBytes)] = utf8Decode(valueBytes);
  }
  return headers;
}

const TE = new TextEncoder();
const TD = new TextDecoder();
function utf8(s: string): Uint8Array {
  return TE.encode(s);
}
function utf8Decode(b: Uint8Array): string {
  return TD.decode(b);
}
